### PR TITLE
Purge shared optimizer

### PR DESCRIPTION
**Pull request by W4ngatang**
Remove mentions of `shared_optimizer` (i.e. always use a single optimizer in multi task learning)

Addresses #597

### CODE REVIEW

File: src/trainer.py [hunk 1]
```diff
@@ -961,18 +946,13 @@ def _validate(self, val_pass, tasks, batch_size, periodic_save=True):
 
     def _get_lr(self):
         """ Get learning rate from the optimizer we're using """
-        if self._g_optimizer is not None:
-            lrs = {"global_lr": self._g_optimizer.param_groups[0]["lr"]}
-        else:
-            lrs = {}
-            for task, task_info in self._task_infos.items():
-                lrs["%s_lr" % task] = task_info["optimizer"].param_groups[0]["lr"]
+        lrs = {"global_lr": self._optimizer.param_groups[0]["lr"]}
```
Comment (sleepinyourhat) [inline src/trainer.py]: Cut this function altogether. It's just making a one-element list for no clear reason now.

File: src/trainer.py [hunk 2]
```diff
@@ -1093,19 +1062,10 @@ def _save_checkpoint(self, training_state, phase="pretrain", new_best=False, tas
...
-        task_states["global"]["optimizer"] = (
-            self._g_optimizer.state_dict() if self._g_optimizer is not None else None
-        )
-        if self._g_scheduler is not None:
-            sched_params = {}
-            task_states["global"]["scheduler"] = sched_params
-        else:
-            task_states["global"]["scheduler"] = None
+        task_states["global"]["optimizer"] = self._optimizer.state_dict()
+        # NOTE(Alex): AllenNLP wrapper doesn't expose scheduler state dict methods
```
Comment (sleepinyourhat) [inline src/trainer.py]: Hrm?

File: src/trainer.py [hunk 3]
```diff
@@ -1204,22 +1164,16 @@ def _restore_checkpoint(self, phase, tasks=None):
...
                 task_state["total_batches_trained"] % self._task_infos[task_name]["n_tr_batches"],
             ):
                 pass
-        if task_states["global"]["optimizer"] is not None:
-            self._g_optimizer.load_state_dict(task_states["global"]["optimizer"])
-        if task_states["global"]["scheduler"] is not None:
-            for param, val in task_states["global"]["scheduler"].items():
-                setattr(self._g_scheduler, param, val)
+        self._optimizer.load_state_dict(task_states["global"]["optimizer"])
+        # NOTE(Alex): AllenNLP wrapper doesn't expose scheduler state dict methods
```
Comment (sleepinyourhat) [inline src/trainer.py]: ?

File: src/trainer.py (hunk 2 above)
Comment (W4ngatang) [reply src/trainer.py]: Just to explain why underneath we have to do `self._scheduler.lr_scheduler`

File: src/trainer.py (hunk 3 above)
Comment (W4ngatang) [reply src/trainer.py]: Just to explain why underneath we have to do `self._scheduler.lr_scheduler`

**Comment by W4ngatang** [comment]
Successfully loaded checkpoint in pretraining and target task training!

**Pull request merged**